docs/optimizer.rst

```diff
@@ -186,3 +186,39 @@ Check out the example below. It contains multiple common subexpressions:
 The expression `a*b` is repeated throughout the function and will be saved to a temporary variable.
 This temporary will be placed wherever there is an expression `a*b`. You can see the pass in action when you compile
 this contract and check the CFG, using `solang --emit cfg`.
+
+
+
+.. _Array-Bound-checks-optimizations:
+
+Array Bound checks optimization
+-------------------------------
+
+Whenever an array access is done, there must be a check for ensuring we are not accessing
+beyond the end of an array. Sometimes, the array length could be known. For example:
+
+.. code-block::
+
+    contract c {
+    	function test() public returns (int[]) {
+    		int[] array = new int[](3);
+    		array[1] = 1;
+    		return array;
+	}
+    }
+    
+In this example we access ``array`` element 1, while the array length is 3. So, no bounds
+checks are necessary and code will more efficient if we do not emit the bounds check in
+the compiled contract.
+
+The array length is tracked in an invisible a temporary variable. This temporary variable
+always kept up to date, when for example a ``.pop()`` or ``.push()`` happens on the array
+or an assignment is done. Then, when the bounds check is done, rather than retrieving the array length from
+the array at runtime, bounds check becomes the constant expression `1 < 3` which is
+always true, so the bounds check is omitted.
+
+This also means that whenever the length of an array accessed using `.length`, this is
+replaced with a constant.
```

Comment:
This whole phrase is unclear and needs rewriting.

Reply:
`As a consequence of this, whenever the length of an array is accessed using '.length', we can replace the corresponding builtin call with a constant.`